### ISSUE TITLE
ImportError: No module named PySimpleGUI

### ISSUE THREAD

**Issue by strk** [wrwetzel/Birdland#3]
It looks like the Install.sh script doesn't deal with installing dependencies as I get an error running `birdland` right after running `./Install.sh`:

```
[strk@oli:/usr/local/src/music/Birdland(main)] birdland 
Traceback (most recent call last):
  File "/home/strk/.local/bin/birdland", line 25, in <module>
    import PySimpleGUI as sg
ModuleNotFoundError: No module named 'PySimpleGUI'
```

Unfortunately I cannot find a package if I search for `PySimpleGUI` in the apt cache ( `apt cache search PySimpleGUI`) so I'm stuck as to how to proceed. Ideas ? The recommended way of using `pip` is discouraged by Debian:

```
[strk@oli:/usr/local/src/music/Birdland(main)] pip install PySimpleGUI
error: externally-managed-environment

× This environment is externally managed
╰─> To install Python packages system-wide, try apt install
    python3-xyz, where xyz is the package you are trying to
    install.
    
    If you wish to install a non-Debian-packaged Python package,
    create a virtual environment using python3 -m venv path/to/venv.
    Then use path/to/venv/bin/python and path/to/venv/bin/pip. Make
    sure you have python3-full installed.
    
    If you wish to install a non-Debian packaged Python application,
    it may be easiest to use pipx install xyz, which will manage a
    virtual environment for you. Make sure you have pipx installed.
    
    See /usr/share/doc/python3.11/README.venv for more information.

note: If you believe this is a mistake, please contact your Python installation or OS distribution provider. You can override this, at the risk of breaking your Python installation or OS, by passing --break-system-packages.
hint: See PEP 668 for the detailed specification.
```

**Comment by strk** [comment]
I'm not a python expert but I've had some nice experience with `pipx install <project>`, how hard would it be to make it possible for birdland as well ?

**Comment by strk** [comment]
I've seen that PySimpleGUI is not opensource software, which explains why it isn't in Debian. Would you consider replacing its use with something that allows to really use your project as free software ?

**Comment by wrwetzel** [comment]
Sandro,

First, an apology for not getting back to you quickly.  I've been busy but that's not a good excuse.

Second, I haven't touched Birdland in a while and I know I left it in a state that was inconvenient to install. I hope to get back to it in the next few months depending on available mindshare.

Now, to your notes. It appears that PySimpleGUI is still available free for non-commercial use but there is a little friction in the process of acquiring it. See: https://www.pysimplegui.com/faq

There is now a free version, FreeSimpleGUI, from another developer who copied the last free release of PySimple
GUI, if I remember correctly. See: https://github.com/spyoungtech/FreeSimpleGUI

I had a quick go at trying it out but bumped up against an install problem with it and didn't put the time into resolving it. 

I explored making Birdland installable with pip, plan to do so, but not immediately. 

I'm afraid that Birdland is so dependent on PySimpleGUI that it would be hard to untangle it from that GUI. I specifically do not have a GUI abstraction layer that would make it easy to replace PySimpleGUI with another UI library. However, to your concern about "free software", I believe the FreeSimpleGUI will satisfy that need.

I think this about addresses your questions. Get back to me if not or you have more questions.
Bill

**Comment by wrwetzel** [comment]
Sandro,

I just downloaded a copy of FreeSimpleGUI from:

     https://github.com/spyoungtech/FreeSimpleGUI/blob/v5.1.0/FreeSimpleGUI/FreeSimpleGUI.py

copied it to the Birdland src folder, commented out the PySimpleGUI include and added:

     import FreeSimpleGUI as sg

and Birdland launched and appeared to run OK. This give you a path to at least explore Birdland.

Bill

**Comment by strk** [comment]
I've tried doing as you suggested and after ./Install.sh and a few
iterations of installing deps and tweaking PYTHONPATH I've got here:

			File "/home/strk/.local/share/birdland/birdland/fb_config.py", line 1160, in report_progress
				self.sg.popup( txt,
				^^^^^^^^^^^^^
			AttributeError: module 'FreeSimpleGUI' has no attribute 'popup'

What about pushing your work in a branch so it's easier to test ?
I have a lot of changes already in the worktree by following your
instructions (used perl to batch-replace  PySimpleGUI with FreeSimpleGUI
everywhere:


		 ReadMe-GitHub.md                                       |   2 +-
		 ReadMe-Tar.md                                          |   2 +-
		 requirements.txt                                       |   2 +-
		 src/Documentation/_build/doctrees/birdland.doctree     | Bin 258209 -> 258247 bytes
		 src/Documentation/_build/doctrees/environment.pickle   | Bin 44733 -> 44737 bytes
		 src/Documentation/_build/html/_sources/birdland.md.txt |  12 ++++++------
		 src/Documentation/_build/html/birdland.html            |  14 +++++++-------
		 src/Documentation/birdland.md                          |  12 ++++++------
		 src/Documentation/birdland.md.bak                      |   4 ++--
		 src/birdland/DevNotes.txt                              |  22 +++++++++++-----------
		 src/birdland/birdland.conf.proto                       |   4 ++--
		 src/birdland/birdland.html                             |   2 +-
		 src/birdland/birdland.py                               |   6 +++---
		 src/birdland/fb_config.py                              |   6 +++---
		 src/birdland/fb_layout.py                              |   6 +++---
		 src/birdland/fb_pdf.py                                 |  10 +++++-----
		 src/birdland/fb_utils.py                               |   8 ++++----
		 17 files changed, 56 insertions(+), 56 deletions(-)

**Comment by wrwetzel** [comment]
Hmmm, I have no idea why popup() is not defined. I'll have to look into that and likely contact the FreeSimpleGUI developers.

"What about pushing your work in a branch so it's easier to test ?" - I'm afraid that I'm a complete beginner when it comes to git. I just learned enough to get this to github. I don't know what you mean by the question. Please explain or, better yet, tell me what to do.

I'll be glad to answer whatever questions you have but I won't have time for actual development work for a while. Several things in the queue ahead of Birdland.

Bill

**Comment by strk** [comment]
How to push your work in a branch:

1. Create the branch (we'll call it "freeSimpleGui"): `git branch freeSimpleGui`
2. Switch to the newly created branch: `git checkout freeSimpleGui`
3. Modify the files as you see needed
4. Commit the changes: `git commit -a`
5. Push the branch to the github repository and indicate it as its upstream (I'm assuming the github remote is called "origin"): `git push --set-upstream origin freeSimpleGui`

**Comment by wrwetzel** [comment]
Sandro,

Several things... 

1) I'm afraid that I dropped the ball trying to follow up on your last comment from almost a year ago. Sorry about that. Lots of reasons why but the bottom line is that birdland got pushed to the back burner and I was giving it no attention.

2) I don't understand the issue with FreeSimpleGUI. I import it in only three files: birdland.py, fb_layout.py, and fb_search.py such that it was not onerous to change from PySimpleGui to FreeSimpleGUI. I've tested birdland.py using FreeSimpleGUI a reasonable amount recently and it appears to be fine. However...

3) PySimpleGUI and FreeSimpleGUI are soon to be a moot point for birdland. In the past couple of months I have completely rewritten the user-interface of birdland using Pyside6, the Qt library for Python. It is nearly feature complete with the exception of the tools for creating new indexes, which I'm not planning to include in the first release. I'm not sure when I will make the first release it but I expect to do so in the next month or two and then add the index-management tools sometime after that. The good news is that because of the portability of Python and Pyside6, birdland also now runs on Windows with essentially no porting effort. I have to learn a bit about packaging for Windows but once that is done I'll release it for Windows, too.

The motivation for a rewrite of the UI was licensing issues with PySimpleGUI and the appearance of a lack of ongoing work with FreeSimpleGUI. Pyside6 turns out to be marginally (at the most) more difficult that PySimpleGUI. In hindsight, I wish I had started with that instead of what I did. 

In rewriting the UI using Pyside6 I also totally restructured the communications between parts of birdland. It now uses Qt Signals and Slots instead of the mess of direct references to UI elements in the prior version. The code is much cleaner, smaller, and easier to maintain.

Stay tuned...

**Comment by BugBaron** [comment]
Very interested to see where this goes, would love to try an alpha build whenever it becomes available! I might hold off on trying to install it for now

**Comment by wrwetzel** [comment]
BugBaron - This project is obsolete, superseded by Birdland-Qt, which is not dependent on PySimpleGUI, and supported on Windows, Linux, macOS. See:

    https://github.com/wrwetzel/Birdland-Qt

and give that a try.

Bill

**Comment by BugBaron** [comment]
Thanks for the quick response!

Just got that started now. It appears that the Windows link is broken, but I managed to install it on Ubuntu through WSL.

FYI in case you want to update the installation notes, I had a couple missing things on a fresh Ubuntu install.
- I got this error when I first tried to start it: `ERROR in main() type: <class 'ImportError'>, value: libOpenGL.so.0: cannot open shared object file: No such file or directory`. To fix that, I just had to install the library manually (`sudo apt install libopengl0`)
- The application crashed when setting things up when a couple expected directories were missing. I had to create the directories `~/.config` and `~/.local/share/applications` manually, since they were not created by default on the Ubuntu installation. (`~/.local/share` was created automatically, but not the final `applications` folder, where the `birdland_qt.desktop` file is created)

The missing folder issues are less likely to be problematic, but it is definitely worth mentioning the libopengl0 installation IMO.

**Comment by wrwetzel** [comment]
All the links at:
https://birdland-qt.wrwetzel.com/
just checked out fine. Where did you see a broken windows link?

The Qt library I'm using, pyside6, includes an abstraction for user directories including the ones you mentioned that were missing. Looks like I'll have to test under Ubuntu. 

Perhaps you can give the windows version a try.

**Comment by BugBaron** [comment]
I am getting a 404 when I try this link: https://birdland-qt.wrwetzel.com/Downloads/birdland_qt_win_installer_2.0.0-beta.1099.exe

Is there another link, or is it something else? Seems strange for only that link to fail if it were something on my end.

**Comment by wrwetzel** [comment]
A big, embarrassing, oops. Nothing on your end, brain fart at my end.
1) I had typos in the windows file name, underscore vs dash.
2) My testing when I claimed all worked was inadequate, just clicked the link, did not actually download.
I just tested the download again, this time more thoroughly, and looks ok.
Please test again.
BTW, are you the Jason Yang who was one of the developers of PySimpleGUI?

**Comment by BugBaron** [comment]
Worked that time, cheers!

I'm afraid that would not be me. I've never heard of a Jason Yang, nor have I touched PySimpleGUI before.

**Comment by wrwetzel** [comment]
Great, sorry for the typo. I appreciate your perseverance. I hope you'll have a chance to install and test it with your own fakebook pdfs. Please get back with any issues. I'll be traveling starting Monday and may take longer to respond but I will keep an eye out for your messages.
